### PR TITLE
feat(taiko-client): change guardian provers  to directly submit proofs instead of contests

### PR DESCRIPTION
**Pull request by YoGhurt111**
## Description
1. Clarify responsibilities between components.
1.1. Event-handler just handles behavior when receiving various events, including submitting proof, submitting a contest, or doing nothing.
1.2. Proof-submitter chooses how to submit proof and whether to use delay.

2. Guardian prover workflow
未命名文件 (2)
未命名文件 (1)
未命名文件

### CODE REVIEW

**Comment by dantaik** [comment]
Gavin, I think there should be a minimum delay before guardian prover submits a proof, to give regular prover a chance to submit valid proofs.

**Comment by YoGhurt111** [comment]
> Gavin, I think when we have zk proofs, there should be a minimum delay before guardian prover submits a proof, to give regular prover a chance to submit valid proofs.

Yes, we did. Either sgx or zk proofs,  guardian prover will not submit proof until the current time is greater than `expiredAt` + at least 1 hour and there was no proof at that time.

In the above guardian prover workflow, I focused on the guardian prover's behavior when receiving events and put the logic about delaying into `packages/taiko-client/prover/proof_submitter/proof_submitter.go`.

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go [hunk 1]
```diff
@@ -134,6 +151,37 @@ func (s *ProofSubmitter) SubmitProof(
...
+			return fmt.Errorf("failed to check if the proving window is expired: %w", err)
+		}
+		// Get a random bumped submission delay, if necessary.
+		submissionDelay, err := s.getRandomBumpedSubmissionDelay(expiredAt)
+		if err != nil {
+			return err
+		}
+		// TODO:  Need to check if there will be performance issues while using simple time.Sleep()
+		time.Sleep(submissionDelay)
+		// Check again.
```
Comment (davidtaikocha) [inline packages/taiko-client/prover/proof_submitter/proof_submitter.go:175]: Shall we only check here?

Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:175]: If we only check here, sometimes we have to still wait until `expiredAt` + at least 1 hour when there's already a proof there. It's not efficient.

Comment (davidtaikocha) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:175]: I see

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go [hunk 2]
```diff
@@ -134,6 +151,37 @@ func (s *ProofSubmitter) SubmitProof(
 		"tier", proofWithHeader.Tier,
 	)
 
+	// Check if we still need to generate a new proof for that block.
+	proofStatus, err := rpc.GetBlockProofStatus(ctx, s.rpc, proofWithHeader.BlockID, s.proverAddress)
+	if err != nil {
+		return err
+	}
+	if proofStatus.IsSubmitted {
```
Comment (davidtaikocha) [inline packages/taiko-client/prover/proof_submitter/proof_submitter.go:159]: If there is a proof submitted, but its invalid, shall we still submit the current proof?

Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:159]: No, it depends on the handle of receiving `TransitionProved` or `TransitionContested` event.

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go [hunk 3]
```diff
@@ -134,6 +151,37 @@ func (s *ProofSubmitter) SubmitProof(
...
+		_, expiredAt, _, err := handler.IsProvingWindowExpired(proofWithHeader.Meta, s.tiers)
+		if err != nil {
+			return fmt.Errorf("failed to check if the proving window is expired: %w", err)
+		}
+		// Get a random bumped submission delay, if necessary.
+		submissionDelay, err := s.getRandomBumpedSubmissionDelay(expiredAt)
+		if err != nil {
+			return err
+		}
+		// TODO:  Need to check if there will be performance issues while using simple time.Sleep()
```
[the remark this answers is not in the document]
Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:173]: ok

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go [hunk 4]
```diff
@@ -170,14 +170,14 @@ func (s *ProofSubmitter) SubmitProof(
 		if err != nil {
 			return err
 		}
-		// TODO:  Need to check if there will be performance issues while using simple time.Sleep()
-		time.Sleep(submissionDelay)
+		delayTimer := time.After(submissionDelay)
```
Comment (davidtaikocha) [inline packages/taiko-client/prover/proof_submitter/proof_submitter.go:173]: After a second thought, seems `time.AfterFunc()` is better?

**Review by davidtaikocha — APPROVED** [review]

File: packages/taiko-client/prover/proof_submitter/proof_submitter.go (hunk 4 above)
Comment (YoGhurt111) [reply packages/taiko-client/prover/proof_submitter/proof_submitter.go:173]: `time.AfterFunc()` needs to pass a function as a parameter, but I just want to continue with the rest of the code.

**Review by mask-pp — APPROVED** [review]

**Pull request merged**